### PR TITLE
Ray/680/topic config overrides

### PR DESCRIPTION
**Pull request by RayRoestenburg**
### What changes were proposed in this pull request?

The topics config should be used in topic actions

### Does this PR introduce any user-facing change?

Fix for not using config file specified topic settings

### How was this patch tested?

GKE

### CODE REVIEW

File: core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala [hunk 1]
```diff
@@ -51,14 +51,58 @@ object TopicActions {
...
-                   labels: CloudflowLabels)(topic: TopicInfo)(implicit ctx: DeploymentContext): CreateOrUpdateAction[ConfigMap] = {
+  def action(secretName: Option[String], namespace: String, labels: CloudflowLabels, topic: TopicInfo)(
+      implicit ctx: DeploymentContext
+  ): Action[ConfigMap] =
+    secretName
+      .map { name =>
+        Action.provided[Secret, ConfigMap](
+          name,
+          namespace, {
+            case Some(secret) =>
```
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala:77]: This can still use some refactoring..

**Review by ennru — COMMENTED** [review]
When actions become more likely to stop deployment when they check the prerequisites, we might need a better way than throwing exceptions to stop it. Would it make sense to accumulate deployment errors/warnings in a more structured way?

File: core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala [hunk 2]
```diff
@@ -51,14 +51,50 @@ object TopicActions {
...
     val labels = CloudflowLabels(newApp)
     val actions =
-      managedTopics.toVector.map(topic => createAction(newApp.namespace, labels)(topic))
+      managedTopics.toVector.map { topic =>
+        // find a config (secret) for the topic, all deployments should have the same configuration for the same topic,
+        // since topics are defined once, deployments refer to the topics by port mappings.
+        // So it is ok to just take the first one found, that uses the topic.
+        val secretName = newApp.spec.deployments
+          .filter(deployment => deployment.portMappings.values.exists(_.id == topic.id))
+          .headOption
```
[the remark this answers is not in the document]
Comment (RayRoestenburg) [reply core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala:60]: sure

File: core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala [hunk 3]
```diff
@@ -51,14 +51,50 @@ object TopicActions {
...
+
+                createAction(namespace, labels, topicInfo)
+              } else {
+                createAction(namespace, labels, topic)
+              }
+            case None => createAction(namespace, labels, topic)
+          }
+        )
+      }
+      .getOrElse(createAction(namespace, labels, topic))
```
[the remark this answers is not in the document]
Comment (RayRoestenburg) [reply core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala:93]: The provided takes `Option[T]`, so not really a lot. Can refactor a little.

File: core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala [hunk 4]
```diff
@@ -131,6 +167,38 @@ object TopicActions {
...
+        getConfig(portMappingsConfig, s"$key.config")
+      }
+  }
+
+  private def toConfig(str: String, secret: Secret): Config =
+    Try(ConfigFactory.parseString(str))
+      .recover {
+        case e =>
+          log.error(
+            s"Detected config in secret '${secret.metadata.name}' that contains invalid configuration data, IGNORING configuration.",
```
[the remark this answers is not in the document]
Comment (RayRoestenburg) [reply core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala]: Improved this.

**Comment by RayRoestenburg** [comment]
Regarding exceptions, right now the operator starts to crash when it encounters particular issues. I'll add a ticket to make this more robust.

File: core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (RayRoestenburg) [reply core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala]: done

File: core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (RayRoestenburg) [reply core/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala]: made a little simpler

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/action/TopicActionsSpec.scala [hunk 5]
```diff
@@ -95,7 +100,15 @@ class TopicActionsSpec
...
-      val configMap0                  = foosAction.asInstanceOf[ResourceAction[_]].resource.asInstanceOf[TopicActions.TopicResource]
+
+      val configMap0 = foosAction
+        .asInstanceOf[ProvidedAction[_, TopicActions.TopicResource]]
+        .getAction(None)
+        .asInstanceOf[ResourceAction[_]]
+        .resource
+        .asInstanceOf[TopicActions.TopicResource]
+
+      //foosAction.asInstanceOf[ResourceAction[_]].resource.asInstanceOf[TopicActions.TopicResource]
```
Comment (ennru) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/action/TopicActionsSpec.scala:111]:
```suggestion
```

**Review by ennru — APPROVED** [review]
LGTM.

**Pull request merged**